docs/source/miscellaneous/doc/adr/0016-use-tickets-encoding-for-transactions.md

```diff
@@ -499,7 +499,159 @@ the worker pool has a finite size and even with the full batching algorithm in t
+
+For a given start timestamp, we want to be able to decide what mechanism to use to find the commit timestamp (or,
+alternatively, if the transaction was aborted).
+
+#### Coordination Service
+
+We introduce a notion of a coordination service, which agrees on values being relevant or correct at a given timestamp.
+The sequence of values being agreed needs to evolve in a *backwards consistent* manner - that is, one should avoid
+changes that may affect behaviour for decisions made at timestamps before the point where one is making one's update.
+More formally, if we read a value for some timestamp ``TS``, then all future values written must then ensure that
+decisions made at ``TS`` would be done in a way consistent with our initial read.
+
+Values provided are valid up to a specific timestamp, called the validity bound. When a transform is applied, a fresh
+timestamp ``F`` is taken, and the new value will be written with a validity bound ``F + C`` for some constant ``C``.
+Going forward, the behaviour for updates up to the previous bound must be preserved (which means decisions would be the
+same whether other service nodes read the new value or not). Nodes must not make decisions based on values read with a
+validity bound less than their timestamp of interest.
+
+```java
+public interface CoordinationService<T> {
+    Optional<ValueAndBound<T>> getValueForTimestamp(long timestamp);
+
+    CheckAndSetResult<ValueAndBound<T>> tryTransformCurrentValue(Function<ValueAndBound<T>, T> transform);
+
+    // ONLY FOR METRICS AND MONITORING - NOT FOR PRODUCTION DECISIONS.
+    Optional<ValueAndBound<T>> getLastKnownLocalValue();
+}
+```
+
+#### Physical Implementation
+
+The coordination service needs to persist state; we've implemented this as a non-transactional table in the AtlasDB
+key-value service called ``_coordination``. Within this table, each sequence of values being agreed on takes up one row.
+The values themselves are stored as dynamic columns with a var-long dynamic column key.
+
+- The value stored at column 0 is a ``SequenceAndBound`` which indicates the sequence number of the currently valid
+  value and the validity bound.
+- Values are stored at column IDs equal to a fresh timestamp at the time they're written.
+
+For example, the Cassandra representation of the coordination table may look as follows.
+
+|  key | column1 | column2 |                                                                  value |
+|-----:|--------:|--------:|-----------------------------------------------------------------------:|
+| 0x6d |    0x00 |      -1 | 0x7b2273657175656e6365223a31333030372c22626f756e64223a353031333030377d |
+| 0x6d |  0xa711 |      -1 |                                     0x7b2276657...3334227d (216 bytes) |
+| 0x6d |  0xb2cf |      -1 |                                     0x7b2276657...3930227d (380 bytes) |
+
+The single ``key`` reflects that these values are all part of the same sequence of values. The column IDs are
```

Comment:
I think it's also useful to concretise the example/tie it back to transactions2.

* what is special about `0x6d` here?, what would it mean for transactions2?
  * 0x6d is the sequence identifier?

It can also do with the above wrt triples perhaps, as I can't make sense of this example whatsoever.

Reply:
`0x6d` is a not-inherently-meaningful sequence identifier which we've picked for AtlasDB usage. Conceptually, if we decided we wanted AtlasDB clients to agree on something else and unrelated to transactions2, they could use a different row in the table.